src/main/java/org/spongepowered/common/data/builder/tileentity/AbstractTileEntityBuilder.java

```
@@ -0,0 +1,94 @@
+
+import com.google.common.base.Preconditions;
+import net.minecraft.block.Block;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.block.BlockType;
+import org.spongepowered.api.block.entity.BlockEntity;
+import org.spongepowered.api.block.entity.BlockEntityType;
+import org.spongepowered.api.data.persistence.*;
+import org.spongepowered.api.world.server.ServerWorld;
+import org.spongepowered.common.accessor.tileentity.TileEntityTypeAccessor;
+import org.spongepowered.common.util.Constants;
+
+import java.util.Optional;
+import java.util.Set;
+
+public abstract class AbstractTileEntityBuilder<T extends BlockEntity> extends AbstractDataBuilder<T> implements DataBuilder<T> {
+    protected AbstractTileEntityBuilder(Class<T> requiredClass, int supportedVersion) {
+        super(requiredClass, supportedVersion);
+    }
+
+    @Override
+    protected Optional<T> buildContent(DataView container) throws InvalidDataException {
+        Preconditions.checkNotNull(container);
+
+        if (!container.contains(Constants.TileEntity.TILE_TYPE, Constants.TileEntity.WORLD, Queries.POSITION_X, Queries.POSITION_Y, Queries.POSITION_Z)) {
+            return Optional.empty();
+        }
+
+        String worldName = container.getString(Constants.TileEntity.WORLD).get();
+        Optional<ServerWorld> worldOptional = Sponge.getGame().getServer().getWorldManager().getWorld(worldName);
+        if (!worldOptional.isPresent()) {
+            throw new InvalidDataException("The provided container references a world that does not exist!");
+        }
+
+        final Optional<BlockEntityType> type = container.getCatalogType(Constants.TileEntity.TILE_TYPE, BlockEntityType.class);
+        if (!type.isPresent()) {
+            // TODO: Double check, do we want to throw an InvalidDataException since the catalog type is not registered at all?
```

I'd say throw InvalidData

---

Same as the world optional handling above. Just `getOrThrow`.